### ISSUE TITLE
Bug: [no-redundant-type-constituents] False positive when using unknown in a union to refine a type parameter in a type guard

### ISSUE THREAD

**Issue by evelynhathaway** [typescript-eslint/typescript-eslint#7325]
### Before You File a Bug Report Please Confirm You Have Done The Following...

- [X] I have tried restarting my IDE and the issue persists.
- [X] I have updated to the latest version of the packages.
- [X] I have [searched for related issues](https://github.com/typescript-eslint/typescript-eslint/issues?q=is%3Aissue+label%3A%22package%3A+eslint-plugin%22) and found none that matched my issue.
- [X] I have [read the FAQ](https://typescript-eslint.io/linting/troubleshooting) and my problem is not listed.

### Playground Link

https://typescript-eslint.io/play/#ts=5.1.6&fileType=.ts&code=MYewdgzgLgBAlhASgU1AJwCYwLwwBQBuAhgDYCuyAXDGWANZggDuYAlNcecvBDCuhgA80NHDABzGAB8YYMgFsARsjTSYEAJ5KQJADQ16jFgD4cpvAFgAUDBicKMAITZcckiWu2AZF-yfbMFAaAA7IIABmdqQOLrgARCCKAFaoUHH%2BtlIyQaERUVw4LjBx4bTAUHDg6TYwrP4%2BTgCCaGhEGgB0CM2tGoTRyHVWrADc1tYA9OMwACLIEHBoyFigGMj6ABZEvOGkENzBIPMVBNwqaCCq4efysiAAtIsYtBhEYFB3Och3oJAVUBRvCDWH7QGCbCAAaWQGgAYrtkAAFQ5wY7cXCCACSYGCZCgAHlkqkYMgAB5QZBgDC8RIpcr6KEaYlkilU9RQUQSNRyJQqNSabQkYyWGqTGAAURaF2oAHJaAxmGBpTAQCcWnBVrxSCRlVB1rzPrwxIF1ggDJUwIEQsh2mKAMoAGTEUAAAgbgKJgu85iQneNGA8ls9Xu9Pt9wNAUQCoECajTUtQsTj8YTymo5UYwLp-HRodQGVmhtQ46nTYncQTabBfABvGAAbQR8AtDIAutR0wqYABfMw8fgXDB4YtQVgwBo5xlG4ejKwTKYACWYMCIMFWJxIIFCqiYIDIJCwPpzJEZRCg5PknsCIBg4TgJON3AQEAowPDsHBDNt613%2B4A6hc6AxcIABU0D3DQUCeSlgxwGBMWxcsU1gUlyUpakkPpaEmVQ1kRDESQZG5ZRVBkflFB0IV-FFFB5BVJYYAAAxkDsWAYs1wH0cIdA3Jh8IfeALx9ejFlAeR5BZU9zWVSJdVOSU0H8YcEwQ5NKwLWwJzzaEC3YZUkJ4GAy1Uoka3rRsjVbdtDE7HtsFMBB%2B0wIckNHccsKnJCZzGKxVmAEgiEWKJVFeDRgKtagQusD9oThEg9iRI44BOPAQrC0J9DiYJzi3II4lc3xUqtdoss3FQghgUVPgMeUWH0DUFiWKKtk-b89wwf80EAkCwOPSCgzeFKwFCq0MpKnKNDyscCqGtLrTGsrGUqq1lyG-RQFoclVCga9ZJgA4nWk-iqvEMgAowfR-VoCAiHCL5xJ5NA7iIYBgDmQ1eDIPYsG2sE4HEdYfX%2B2BdQQIA&eslintrc=N4KABGBEBOCuA2BTAzpAXGUEKQAIBcBPABxQGNoBLY-AWhXkoDt8B6Jge1ukQBNYmvAIYtaRUrTIcmyfJXyxELVBkiJo0DtEgAacNjzjyVGvWSMW7LgORCAZoloBbRE4BG62kLJkUKqOqa2voAviAhQA&tsconfig=N4KABGBEDGD2C2AHAlgGwKYCcDyiAuysAdgM6QBcYoEEkJemy0eFYDAruuGAL4g9A&tokens=false

### Repro Code

```typescript
const isRecord = (value: unknown): value is Record<string | number | symbol, unknown> => (
  value !== null
  && (
    typeof value === "object"
    || typeof value === "function"
  )
  && !Array.isArray(value)
);

// Desired code, has false positive error from no-redundant-type-constituents
const hasKeyFalsePositive = <InputObject extends object, Key extends string | number | symbol>(
  // Error: 'unknown' overrides all other types in this union type.ESLint@typescript-eslint/no-redundant-type-constituents
  object: InputObject | unknown,
  key: Key,
): object is InputObject & { [P in Key]: unknown } => isRecord(object) && key in object;

// How a developer would likely attempt to fix the issue
const hasKeyShouldWorkIfTrulyRedundant = <InputObject extends object, Key extends string | number | symbol>(
  // Removed `| unknown` union, following the implied recommendation of the error
  object: InputObject,
  key: Key,
): object is InputObject & { [P in Key]: unknown } => isRecord(object) && key in object;


declare var anyType: any
hasKeyFalsePositive(anyType, "property") && anyType.property // type unknown, desired
hasKeyShouldWorkIfTrulyRedundant(anyType, "property") && anyType.property // type any, counter to the point of the type guard, no-unsafe-member-access is used to highlight this
```

### ESLint Config

```javascript
module.exports = {
  "rules": {
    "@typescript-eslint/no-redundant-type-constituents": "error",
    "@typescript-eslint/no-unsafe-member-access": "error"
  }
}
```

### tsconfig

```jsonc
{
  "compilerOptions": {
    "strict": true
  }
}
```

### Expected Result

The first `hasKey()` variation, `hasKeyFalsePositive()` should not get a linting error, or at least the error message should indicate that this is *sometimes* redundant, not always.

### Actual Result

`hasKeyFalsePositive()` causes a linting error with an error message that would lead developers to produce a "fix" that makes the linting error go away, but cause type errors on any dependent code.

**Comment by bradzacher** [comment]
I think you've misunderstood the lint rule.
The rule isn't saying "remove the unknown because it overrides everything" - instead it's saying "remove everything else because it's redundant and overriden by the unknown".

I.e. The fix is not

```diff
-object: InputObject | unknown,
+object: InputObject,
```

Instead the fix is

```diff
-object: InputObject | unknown,
+object: unknown,
```

> hasKeyFalsePositive() should not get a linting error, or at least the error message should indicate that this is sometimes redundant, not always.

The rule is correct though - the other constituents are all redundant in the presence of `unknown` - there's never a situation where that isn't true.

**Comment by evelynhathaway** [comment]
Woah. Yes, you totally are correct. I am not sure why we have the `InputObject` in the type guard since it behaves just like an `as InputObject`, which only adds looseness since TypeScript does not need it for type narrowing.

Thanks so much!
